### ISSUE TITLE
Datums / Datum Sets

### ISSUE THREAD

**Issue by brycemcanally** [pachyderm/pachyderm#3703]
I think I am at a point with the storage layer where I/we need to start thinking about the information we will store for datums / datum sets. I will go ahead and layout the current storage for datums in Pachyderm and identify the problems. Then, I will just talk about some challenges/ideas that have come up over the course of the 1.8 improvements and the storage layer design process.

**Current Model:**

Each datum is downloaded, processed, and uploaded independently of other datums. The results of each datum are stored as two objects in object storage (hashtree and data). Each datum is identified by a hash created from information about the inputs and pipeline. We upload the hashtree to object storage with a tag consisting of this hash. These tags are used when determining whether we can skip a datum, which provides processing incrementality. The merge process prior to 1.8 would always merge these hashtrees from scratch for every job, which for obvious reasons was not providing the desired incrementality for workloads with expensive merges or lots of small jobs. With 1.8, we improved the merge process and implemented an optimization that would re-use the prior hashtree in cases where the workload was additive (only new datums added). This is nice, but it still means we have to re-merge from scratch when a datum needs to be removed from the parent commit's hashtree.

**Today:**

We are intending to address this shortcoming by adding information in the storage layer about what datums contributed what content to what files. The obvious solution seems to be a reverse index, a mapping from some datum identifier to a set of bytes in a file. Having this mapping should in theory make it possible to remove old datums and insert new ones. The question though is how do we represent a datum in the reverse index and quickly find which bytes need to be removed (the adding bytes part is easier because it is just merging in the new stuff). There is also a scalability challenge with needing to go to object storage for every datum during every job to check if it can be skipped or not (there is currently an optimization where the datums that can be skipped are serialized into an object, but it only applies to datums that were processed in a successful job).

**Challenges/Ideas:**

- The reverse index can get very large for certain workloads. In the worst case, n x m entries where n is number of datums and m is number of output files. So, if we want to store a set of entries for each file that contains enough information to know what bytes a set of datums contributed, then we need to be very careful about the size of each entry. What I have proposed as of now is that we have an ordered list of entries that each contain an id and size of type int64. This means that each entry is 16 bytes, which means that we could handle a lot of them in most cases. The challenge with this is figuring out a mapping from the id back to a datum. I think there might be an alternative model for this, but I don't know what it is yet.
- Figuring out if we want a consistent ordering of the output, and if so, what that ordering is. If we want output data to have the same order regardless of the order in which its inputs were created (multiple vs single commit, different orders for multiple commits), then I think our reverse index needs to capture order information. What has been proposed up to this point is that the output order should correspond with the lexicographical order of the input datums (which is a little unclear for datums that consist of multiple inputs).
- We don't want to store two objects per datum if we have a lot of datums with small output, so we need to figure out some way to batch them. It seems to make sense to group them into datum sets (what we call chunks now), but what that means in terms of incremental processing and merging is pretty vague at this point. The way I was thinking of it is that we could think of datum sets as sort of the checkpoints in a job, such that a user could effectively control batching of their data (if they know they have a lot of quick processing/small output datums) through configuring the datum sets. Users with workloads with long processing/large datums would just decrease the number of datums in a datum set.

**Comment by jdoliner** [comment]
> The challenge with this is figuring out a mapping from the id back to a datum. I think there might be an alternative model for this, but I don't know what it is yet.

I was thinking that we'd identify the datums with a hash of the filenames, we could probably pack those into an int64, not sure if that would be as collision resistant as we need, but it might get close.

> Figuring out if we want a consistent ordering of the output, and if so, what that ordering is.

I think we want this for a few reasons, 1) I think it's a good feature, it makes support for things like timeseries workloads a lot nicer. But possibly more important is 2) I think we need this for our deduplication to be as robust as we want it to be. If things aren't merging in a consistent order we're going to start getting into situations where we're storing multiple copies of the same data ordered differently.

>  (which is a little unclear for datums that consist of multiple inputs).

We probably just order based on the order the inputs appear. I.e. compare first inputs first, if they're the same compare second inputs, etc. Or does that not work?

> but what that means in terms of incremental processing and merging is pretty vague 

It seems like the reverse index helps us quite a bit with this, since we no longer need to worry about having an output that we want to use except for having data from a few datums that have changed. We can just merge and filter out the datums we don't want.

**Comment by brycemcanally** [comment]
> I was thinking that we'd identify the datums with a hash of the filenames, we could probably pack those into an int64, not sure if that would be as collision resistant as we need, but it might get close.

After talking with Matt a bit about this, I think I am convinced that we do not want to try and fit these identifiers into an int64. We probably just want to go with a string for the file path (or some concatenation for multiple inputs) for the datum. This can still get pretty large for map-reduce jobs, but I think it realistically will never approach worst-case performance (n x m) and we can also include information about this under some best practices guide that will talk about batching / fitting more work into each datum such that the number of datums can remain smallish while still mapping to a large number of output files.

> I think we want this for a few reasons, 1) I think it's a good feature, it makes support for things like timeseries workloads a lot nicer. But possibly more important is 2) I think we need this for our deduplication to be as robust as we want it to be. If things aren't merging in a consistent order we're going to start getting into situations where we're storing multiple copies of the same data ordered differently.

k

> It seems like the reverse index helps us quite a bit with this, since we no longer need to worry about having an output that we want to use except for having data from a few datums that have changed. We can just merge and filter out the datums we don't want.

The part that is vague is how to efficiently figure out what needs to be removed and added, not the actual adding/removing process.

**Comment by brycemcanally** [comment]
**Thoughts on datum sets and propagating diffs / incrementality**

**New Stuff:**
- Each commit will have an associated diff file set.
- Stats will be replaced with a datum file set (which subsumes the stats functionality).

**Notes about terminology:**
- Full file sets are full snapshots of a file system at a particular repo and branch.
- Base file sets are the full file sets (normal or datum) associated with the parent commit of the commit being processed.
- Datum file sets are associated with output commits.
- In general, some of this terminology could be improved, but I think it is workable right now.

**Diff File Set:**
- When a commit is finished, the data and operations associated with that commit will be recorded in the format of a diff file set. The diff file set is formed by merging all of the data and operations for the commit. This diff file set will be merged into the base file set to create the full file set for the finished commit. Both the diff and full file set will have semantic objects in object storage.

**Datum File Set:**
- When a job is finished, the data and operations associated with each of the processed datums will be merged into the datum file set. The datum file set is structured in such a way to provide a direct mapping from a datum to its input and output as well as some additional metadata.

*Format:*
```
/datum-path-1/meta
  	     /pfs/in
		 /out
/datum-path-2/meta
	     /pfs/in
		 /out
/datum-path-3/meta
	     /pfs/in
		 /out
```

- The meta file will store additional metadata about the datum (hash, stats, etc.). It may make sense for this to be a protocol buffer if we expect this to contain low volume data mainly because it would be easy to add new metadata later.

**Computing datums to process/merge:**

The datums associated with a job are the paths that match a glob pattern in the full file set associated with the commit that is used as input (keeping it simple right now with one input). A simple way to compute these paths is to read the input commit full file set and keep track of which paths match. If instead we use the input commit diff file set, the operation would be matching the glob pattern to a prefix of the paths in the diff file set to determine which datums need to be processed (sometimes just removed) for the job. This information is all we will need in terms of creating datum sets.

**How this could play out in a job:**

*Processing:*

- The master creates datum sets based on ranges of datums found in the diff file set.
- Workers claim these datum sets and do the processing (download datums from the input commit full file set) and merge the results into the normal and datum file set for the datum set. The normal file set for the datum set will be a diff for the datums within the datum set. There are three particular cases that need to be handled here:
	- Datum added: Worker processes/merges the datum and marks all of the output files in the normal and datum file set with the append operation. 
		- This case applies when a datum is found in the input commit full file set, but not the base datum file set.
	- Datum modified: Similar to when a datum is added, but also reads the output from the base datum file set and adds the deletion operations for the files that were removed.  
		- This case applies when a datum is found in the input commit full file set, and the base datum file set (and the hash is different).
	- Datum removed: Just the latter part of the datum modified case.
		- This case applies when a datum has been deleted in the input commit full file set, and exists in the base datum file set.

*Merging:*

- As datums are processed, their results will be merged into a normal and datum file set associated with their respective datum sets.
- At the end of the job, each datum set's normal file set will be merged into a diff file set for the output commit (this will be used downstream similarly to how we use the diff file set associated with the input commit). Then, the diff file set and datum file sets will be merged into the respective base file sets (it may make sense to also have a diff datum file set, but I am a little worried that we might already be pushing the storage a bit, particularly on workloads where there will not be much deduplication between the normal and datum file sets).

**Random Notes:**

- Directory hashes will be computed at read time (for jobs we will take advantage of the diffs to not have to compute hashes for all of the directories). Also, directory size will be computed at read time (probably makes sense to have directory size not be computed on basic list operations).
- We will store a datum's hash in the meta file. This will be used by the following jobs to determine if a datum needs to get processed or not. It will be updated when a datum gets modified.

**Comment by brycemcanally** [comment]
**Side note about the above post:**

This model might actually let us bring back intra-datum incrementality in a sensible way. This model gives us a clear notion of the previous version of a datum (specifically just the immediately previous). This of course hinges on the datum path being the identifier for a datum across jobs.

**Comment by jdoliner** [comment]
A few thoughts and clarifications:

There's nothing special about the format of diff file sets, they're simply file sets that are understood to have a different meaning, is that correct?

This leaves out one of the trickiest parts of the processing step which is how you find datums which weren't added in this commit but weren't processed in the commit in which they were added due to processing failures (or the pipeline not existing when the commit was created). This seems like it's going to be pretty similar to what's needed to cross new datums in one input with old datums in another.

I'm not sure I understand what you're talking about with `GlobFile` needing to support a notion of matching prefixes, don't globs already allow you to match prefixes?

> I think we should try to make datum iteration a streaming operation instead of reading in the full set of file infos. For crosses, the iterators should cache file infos themselves.

This is definitely necessary, in fact it needs to be a paginated streaming operation for it to work, so that works don't have to start at the beginning of the stream each time.

**Comment by brycemcanally** [comment]
> There's nothing special about the format of diff file sets, they're simply file sets that are understood to have a different meaning, is that correct?

Correct

> This leaves out one of the trickiest parts of the processing step which is how you find datums which weren't added in this commit but weren't processed in the commit in which they were added due to processing failures (or the pipeline not existing when the commit was created). This seems like it's going to be pretty similar to what's needed to cross new datums in one input with old datums in another.

I think this is conceptually pretty simple. Basically, for each of the pfs inputs you need to walk backwards in the commit chain starting at the input commit until you reach the commit that matches the one associated with the most recent successful job. Then, make a merge call on all of the diff file sets associated with the commits encountered. If you mean tricky in the sense that deciding how far to go back before just using the full file set is tricky, then that may be true. The pipeline not existing when the commit was created is an easy one because the first job should just read the full file set from the input commits anyways.

> I'm not sure I understand what you're talking about with GlobFile needing to support a notion of matching prefixes, don't globs already allow you to match prefixes?

This is actually not necessary. I was thinking we could be in a situation where we would have a file path in a diff file set that was under a directory without the corresponding directory entry. If that was the case, you would have to match the directory in the file path (if the directory matched the glob pattern). So, if you just had the file path `/d1/f1` and your glob pattern was `/*`, then you would need to match the prefix `/d1`.

> This is definitely necessary, in fact it needs to be a paginated streaming operation for it to work, so that works don't have to start at the beginning of the stream each time.

I assume you mean workers? I was thinking they would operate on path ranges, not some arbitrary index. A worker's iterator would index into the diff file set based on this range. Although, I still see the use case for pagination in terms of Dashboard stuff.

**Comment by jdoliner** [comment]
> I think this is conceptually pretty simple. Basically, for each of the pfs inputs you need to walk backwards in the commit chain starting at the input commit until you reach the commit that matches the one associated with the most recent successful job. Then, make a merge call on all of the diff file sets associated with the commits encountered. If you mean tricky in the sense that deciding how far to go back before just using the full file set is tricky, then that may be true. The pipeline not existing when the commit was created is an easy one because the first job should just read the full file set from the input commits anyways.

This makes sense, thanks for clarifying. Is there a compelling reason to merge then iterate rather than iterating the different diffs in parallel?

> I assume you mean workers? I was thinking they would operate on path ranges, not some arbitrary index. A worker's iterator would index into the diff file set based on this range. Although, I still see the use case for pagination in terms of Dashboard stuff.

Yup, definitely meant workers. Operating on path ranges makes sense, I think it's likely still worthwhile to have numerical pagination, since it's useful for dash and other clients as you mention. And also probably would be useful for computing those ranges.

**Comment by jdoliner** [comment]
Also worth mentioning, I think this basically brings us full circle on greedy vs. lazy merges. Writing out a diff file set is effectively deferring the merge until a later point. But it has a lot of added value here because in a lot of cases you don't want to read the full dataset, you just want to read the diff. One thing that I would like to see clarified a bit, that you're probably already working on, is what does the process for starting a merge look like. I think this is probably going to be similar to, or maybe even the same as the "symbolic merge" that we had talked about before.

**Comment by brycemcanally** [comment]
> Also worth mentioning, I think this basically brings us full circle on greedy vs. lazy merges. Writing out a diff file set is effectively deferring the merge until a later point. But it has a lot of added value here because in a lot of cases you don't want to read the full dataset, you just want to read the diff. 

I don't think having a notion of a diff file set is quite the same as the lazy merge we originally had in mind. We are still merging the full file set eagerly, but we are keeping around a diff for each commit. What those diffs mean and how they are used will depend on the pipelines that have the branch as input. Merging them only becomes relevant when failed/killed jobs are involved. The primary use I had in mind for the diff file set was indicating what datums to process and what needs to be deleted, then using that information to extract the full information from the full file set (although there are some cases where this is unnecessary).

> One thing that I would like to see clarified a bit, that you're probably already working on, is what does the process for starting a merge look like. I think this is probably going to be similar to, or maybe even the same as the "symbolic merge" that we had talked about before.

The symbolic merge is tricky because we are operating on object storage. I have been thinking in terms of just semantic paths, with the common case being to merge everything under a semantic prefix (everything under a commit prefix for example).

**Comment by jdoliner** [comment]
I think it's the same as a lazy merge in the sense that writing out what to merge and the process of merging are separate, and thus can in theory be delayed as long as we'd like to. In practice the plan right now is to start that merge process immediately, but I suspect that will change at some point soon and I think we're pretty well setup to do so.

> the common case being to merge everything under a semantic prefix (everything under a commit prefix for example).

It seems like just recording that information, what you want to merge, in a database or etcd would effectively be a symbolic merge. Since it records what to merge without doing the merge. And I think that's basically what we had talked about since you need to have some method of communicating to workers what you want to merge and keep track of their progress on it.
